services/deployment-agent/src/simcore_service_deployment_agent/git_url_watcher.py

```diff
@@ -16,53 +17,74 @@
 NUMBER_OF_ATTEMPS = 5
 MAX_TIME_TO_WAIT_S = 10
 
-
 async def _git_clone_repo(repository: URL, directory: Path, branch: str, username: str = None, password: str = None):
-    cmd = "git clone -n {repository} --depth 1 {directory} --single-branch --branch {branch}".format(
-        repository=URL(repository).with_user(username).with_password(password), directory=directory, branch=branch)
-    await run_cmd_line(cmd)
-
-
-async def _git_checkout_files(directory: Path, paths: List[Path]):
-    cmd = "cd {directory} && git checkout HEAD {path}".format(directory=directory,
-                                                              path=" ".join(paths))
+    cmd = f"git clone -n {URL(repository).with_user(username).with_password(password)} --depth 1 {directory} --single-branch --branch {branch}"
     await run_cmd_line(cmd)
 
 
-async def _git_checkout_file(directory: Path, file_path: Path):
-    await _git_checkout_files(directory, [file_path])
+async def _git_get_current_sha(directory: Path) -> str:
+    cmd = f"cd {directory} && git rev-parse --short HEAD"
+    sha = await run_cmd_line(cmd)
+    return sha.strip("\n")
 
+async def _git_clean_repo(directory: Path):
+    cmd = f"cd {directory} && git clean -dxf"
+    await run_cmd_line(cmd)
 
-async def _git_checkout_repo(directory: Path):
-    cmd = "cd {directory} && git checkout HEAD".format(directory=directory)
+async def _git_checkout_files(directory: Path, paths: List[Path], tag: str = None):
+    if not tag:
+        tag = "HEAD"
+    cmd = f"cd {directory} && git checkout {tag} {' '.join(paths)}"
     await run_cmd_line(cmd)
 
+async def _git_checkout_repo(directory: Path, tag: str = None):
+    await _git_checkout_files(directory, [], tag)
 
 async def _git_pull_files(directory: Path, paths: List[Path]):
-    cmd = "cd {directory} && git checkout FETCH_HEAD {path}".format(directory=directory,
-                                                                    path=" ".join(paths))
+    cmd = f"cd {directory} && git checkout FETCH_HEAD {' '.join(paths)}"
     await run_cmd_line(cmd)
 
 
 async def _git_pull(directory: Path):
-    cmd = "cd {directory} && git pull".format(directory=directory)
+    cmd = f"cd {directory} && git pull"
     await run_cmd_line(cmd)
 
-
 async def _git_fetch(directory: Path):
-    cmd = "cd {directory} && git fetch --prune".format(directory=directory)
+    cmd = f"cd {directory} && git fetch --prune --tags"
     await run_cmd_line(cmd)
 
+async def _git_get_latest_matching_tag(directory: Path, regexp: str) -> List[str]:
+    cmd = f"cd {directory} && git --no-pager describe --tags --always `git rev-list --tags`"
+    all_tags = await run_cmd_line(cmd)
+    if not all_tags:
+        return []
+    list_tags = all_tags.split("\n")
+    reg_object = re.compile(regexp)
+    for tag in list_tags:
+        if reg_object.fullmatch(tag):
+            return tag
+    return ""

```

Comment:
Why not returning ``None``?
Actually the returned type does not match ``List[str]`` ... do you mean ``Optional[None, str]. Same applies for `` if not all_tags: return [] `` above.

Reply:
again good catch. but BTW it needs to be used as Option[str] = Union[None, str].